Data/models/History.swift

```diff
@@ -104,60 +104,26 @@ public class History: NSManagedObject, WebsitePresentable {
     }
 
     class func getExisting(_ url: URL, context: NSManagedObjectContext) -> History? {
-        assert(!Thread.isMainThread)
-
-        let fetchRequest = NSFetchRequest<NSFetchRequestResult>()
-        fetchRequest.entity = History.entity(context)
-        fetchRequest.predicate = NSPredicate(format: "url == %@", url.absoluteString)
-        var result: History? = nil
-        do {
-            let results = try context.fetch(fetchRequest) as? [History]
-            if let item = results?.first {
-                result = item
-            }
-        } catch {
-            let fetchError = error as NSError
-            print(fetchError)
-        }
-        return result
+        let urlKeyPath = #keyPath(History.url)
+        let predicate = NSPredicate(format: "\(urlKeyPath) == %@", url.absoluteString)
+        
+        return first(where: predicate)
     }
 
     class func frecencyQuery(_ context: NSManagedObjectContext, containing:String? = nil) -> [History] {
-        assert(!Thread.isMainThread)
+        let urlKeyPath = #keyPath(History.url)
+        let visitedOnKeyPath = #keyPath(History.visitedOn) 
 
-        let fetchRequest = NSFetchRequest<NSFetchRequestResult>()
-        fetchRequest.fetchLimit = 100
-        fetchRequest.entity = History.entity(context)
-        
-        var predicate = NSPredicate(format: "visitedOn > %@", History.ThisWeek as CVarArg)
+        var predicate = NSPredicate(format: "\(visitedOnKeyPath) > %@", History.ThisWeek as CVarArg)
         if let query = containing {
-            predicate = NSPredicate(format: predicate.predicateFormat + " AND url CONTAINS %@", query)
-        }
-        
-        fetchRequest.predicate = predicate
-
-        do {
-            if let results = try context.fetch(fetchRequest) as? [History] {
-                return results
-            }
-        } catch {
-            let fetchError = error as NSError
-            print(fetchError)
+            predicate = NSPredicate(format: predicate.predicateFormat + " AND \(urlKeyPath) CONTAINS %@", query)
```

Comment:
NSPredicate(format: predicate.predicateFormat + " AND \(urlKeyPath) CONTAINS \(query)")